pkg/cmd/roachtest/main_test.go

```diff
@@ -98,3 +105,136 @@ func TestSampleSpecs(t *testing.T) {
+		return db, err
+	}
+	t.Run("expect CategoriseTests to fail which causes fall back to run all tests", func(t *testing.T) {
+		// The failure of the CategoriseTests does not cause any failure. It falls back to run all the tests in the spec
+		db, mock, err = sqlmock.New()
+		require.Nil(t, err)
+		specs, _ := getMockData()
+		mock.ExpectPrepare(regexp.QuoteMeta(testselector.PreparedQuery)).WillReturnError(fmt.Errorf("failed to prepare"))
+		updateSpecForSelectiveTests(ctx, specs)
+		for _, s := range specs {
+			if !strings.Contains(s.Name, "skipped") {
+				require.Empty(t, s.Skip)
+			} else {
+				require.Equal(t, "test spec skip", s.Skip)
+			}
+		}
+	})
+	t.Run("expect no failure", func(t *testing.T) {
+		db, mock, err = sqlmock.New()
+		require.Nil(t, err)
+		specs, rows := getMockData()
+		mock.ExpectPrepare(regexp.QuoteMeta(testselector.PreparedQuery))
+		mock.ExpectQuery(regexp.QuoteMeta(testselector.PreparedQuery)).WillReturnRows(rows)
+		oldSuite := roachtestflags.Suite
+		roachtestflags.Suite = registry.Nightly
+		defer func() {
+			roachtestflags.Suite = oldSuite
+		}()
+		roachtestflags.SuccessfulTestsSelectPct = 0.30
+		specsLengthBefore := len(specs)
+		updateSpecForSelectiveTests(ctx, specs)
+		require.Equal(t, specsLengthBefore, len(specs))
+		for _, s := range specs {
+			if strings.Contains(s.Name, "success_skip_selector") {
+				require.Equal(t, "test selector", s.Skip, s.Name)
+				require.Equal(t, "test skipped because it is stable and selective-tests is set.", s.SkipDetails, s.Name)
+			} else if strings.Contains(s.Name, "skipped") {
+				require.Equal(t, "test spec skip", s.Skip, s.Name)
+				require.Equal(t, "test spec skip test", s.SkipDetails, s.Name)
+			} else {
+				require.Empty(t, s.Skip, s.Name)
+			}
+			if strings.Contains(s.Name, "_preempted") {
+				require.True(t, s.IsLastFailurePreempt(), s.Name)
+			} else {
+				require.False(t, s.IsLastFailurePreempt(), s.Name)
+			}
+		}
+	})
+}
+
+// getMockData returns the mock data as:
+// 1. List of test specs.
+// 2. List of rows to be returned by snowflake.
+// The data follows a convention:
+// all tests which contains "_skipped" are skipped in the spec
+// all tests which contains "_selected" are selected="yes"
+// all tests which contains "_preempted" are last_failure_is_preempt="yes"
+// all tests which contains "_skip_selector" are skipped based on the criteria
+// all tests which contains "_missing" are missing in the test spec
+// all tests which contains "_success" are selected="no"
+// all tests which contains "new_test" are present in specs, but missing in snowflake rows
+// all tests which contains "_randomized" are set in spec as Randomized=true
+// Other than the above conventions, the name of the test is randomly put to represent random names for tests.
+// the test names are also shuffled to ensure that the test is more close to real data.
+func getMockData() ([]registry.TestSpec, *sqlmock.Rows) {
```

Comment:
nit: maybe `getTestSelectionMockData`? Since this is in `main_test.go` lets make it specific.